### PR TITLE
feature: add first metamodel support

### PR DESCRIPTION
**Pull request by monperrus**
first step for #1183 

WDYT?

### CODE REVIEW

**Comment by pvojtechovsky** [comment]
It is on good way! May be I will have Xmas sooner? ;-)

Some ideas:

May be it might return Set instead of Collection.

The returned references have assigned different Factory instance, may be we should add new type of spoon object under Factory, called  `MetaModel`, which would provide such information (and which would have access to the correct parent factory.)

Or return Set<String> with qualified names of interfaces? It is faster to search in it.

File: src/main/java/spoon/Metamodel.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		result.add(factory.Type().get(spoon.reflect.reference.CtIntersectionTypeReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtLocalVariableReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtPackageReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtParameterReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtTypeParameterReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtTypeReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtUnboundVariableReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtVariableReference.class));
+		result.add(factory.Type().get(spoon.reflect.reference.CtWildcardReference.class));
```
Comment (surli) [inline src/main/java/spoon/Metamodel.java]: Why do you do this by hand instead of retrieving all classes of interesting packages by reflection?

Comment (tdurieux) [inline src/main/java/spoon/Metamodel.java]: Do you know the performance of reflection?

**Comment by monperrus** [comment]
> May be it might return Set instead of Collection.

Agree, done.

> The returned references have assigned different Factory instance, 

Yes, but the `equals` contract works across factories.

> may be we should add new type of spoon object under Factory, called MetaModel, which would provide such information (and which would have access to the correct parent factory.)

Let's start with a stateless design first (stateless is better), we'll modify it later if needed. 

> Or return Set with qualified names of interfaces? It is faster to search in it.

I would prefer domain-specific types over strings

**Pull request merged**